README
DSWS (Dylan's simple web server) - This name is not final and may change, same with everything on this project. 
ALPHA v0.3

DSWS is a simple python web server designed to be deployed within seconds of being downloaded.

Current features:
    - Fast, stable webserver.
    - Decent scripting support.

Feature roadmap:
    - Sessions and post / get variables accessible by scripting.
    - Caching system
    - Better error logging / handling
    - Clean up code alot more!

Update notes:
    - ALPHA v0.1 (~ September 2018)
        - The simple webserver is now working.
        - Webpages are being loaded from file.

    - ALPHA v0.2 (26th May 2019)
        - Resumed development after huge initial break.
        - Cleaned up code massively.
        - Switched to being single threaded (will go for an asynchronous architecture in next update.)
        - Added scripting support.
        - Added config file.
        - Started working on logging.

    - ALPHA v0.3 ()
        - Seriously cleaned up code and added tonnes of comments.
        - Server is now fully asynchronous!
        - Scripting works a lot better now. (formatting issues were present)
        - Added caching support.
        - Fairly close to a full release version now, probably a few more alpha versions.

CHANGELOG
- ALPHA v0.1 (~ September 2018)
    - The simple webserver is now working.
    - Webpages are being loaded from file.

- ALPHA v0.2 (26th May 2019)
    - Resumed development after huge initial break.
    - Cleaned up code massively.
    - Switched to being single threaded (too many reasons to list...)
    - Added scripting support.
    - Added config file.
    - Started working on logging.

simplepythonwebserver.py
import socket
import time
import threading

# Dylan's simple python web server, I chose to write a webserver because networking is really easy in python and webservers are incredibly simple.
# I could eventually accept parameters in to the request (this would let me create dynamic web pages. So things like logins, sessions etc.)

# The run down of how my webserver works :
# 1) Server listens for client.
# 2) Client connects and sends over HTTP header. This contains things such as which page they want to visit, their user agent etc.
# 3) The server recieves the clients http header and parses it. (extracts the information from the request)
# 4) The server sends over the page that the client has requested.
# 5) Connection terminates and client is left with a fully loaded webpage on their side.

bindIp = '127.0.0.1'
httpPort = 80
bufferSize = 128 * 1024
rootFolder = "www"

def getFile(file):
        try:
                fileHandle = open(file, "r")
                fileContents = fileHandle.read()
                fileHandle.close()
                return fileContents
        except:
                return False

def serverMessage(message, colour):     #will add more formatting later on
        return "<h3><font color='{col}'>{msg}</font></h3>".format(msg=message, col=colour)

def genHeaders(responseCode):
        header = ''
        if responseCode == 200:
                header += "HTTP/1.1 200 OK\n"
        elif responseCode == 404:
                header += "HTTP/1.1 404 Not Found\n"
        elif responseCode == 501:
                header += "HTTP/1.1 501 Not Implemented\n" 

        timeNow = time.strftime("%a, %d %b %Y %H:%M:%S", time.localtime())
        header += 'Date: {now}\n'.format(now=timeNow)
        header += 'Server: Dylans-Python-Webserver\n'
        header += 'Connection: close\n\n'
        return header

def parseHeader(header, param):
        header_array = header.split("\r\n")
        if param == "method":
                return header_array[0].split(" ")[0]
        elif param == "page":
                return header_array[0].split(" ")[1]
        else:
                for header_line in header_array:
                        param_split = header_line.split(":")
                        if param_split[0].lower() == param:
                                return param_split[1]

def handleClient(clientConnection):
        timerStart = time.time() * 1000
        data = clientConnection.recv(bufferSize).decode()
        if not data:
            return
        
        responseHeader = ""
        responseData = ""

        requestMethod = parseHeader(data, "method")
        
        if requestMethod == "GET":
                responseHeader = genHeaders(200)
                requestedPage = parseHeader(data, "page")
                print("A client requested : {req}\r\n".format(req=requestedPage))
                if requestedPage == "/":
                        responseData = getFile(rootFolder + "/index.html")
                else:
                        responseData = getFile(rootFolder + requestedPage)
        else:
                responseHeader = genHeaders(501)
                responseData = serverMessage("Unknown request method '{reqmethod}'".format(reqmethod=requestMethod), "red")      

        if responseData == False:
                responseHeader = genHeaders(404)
                responseData = serverMessage("Error 404, the page you are looking for cannot be found.", "red")

        response = responseHeader.encode()
        timeTook = round((time.time() * 1000)-timerStart)
        responseData += "\n\nIt took a incredible {genin}ms to generate this page".format(genin=timeTook)
        response += responseData.encode()
        
        conn.send(response)
        conn.close()

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind((bindIp, httpPort))
s.listen(5)

while True:
        conn, addr = s.accept()
        threading.Thread(target=handleClient,args=(conn,)).start()


server.py
import socket
import time
import json
import os
import re

class webserver:
    def __init__(self, configLocation):
        try:
            with open(configLocation) as cfg:
                self.config = json.load(cfg)
        except:
            print(f"Could not load the config, are you sure {configLocation} exists?")
            exit()
        self.parsedHeaders = {}
        self.cache = {}
        self.s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.s.bind(("localhost", self.config["port"]))
        self.s.listen(5)

    def handleClient(self):
        bufSize = self.config["buffersize"]*1024
        data = self.conn.recv(bufSize).decode(errors="ignore")
        if not data:
            return
        self.parseHeaders(data)
        if self.parsedHeaders["method"] == "GET":
            response = self.genHeaders(200)
            route = self.getRoute()
            if not route:
                self.conn.close()
                return
            response += route
            #self.logMessage("Handled {} : {}".format(self.addr, requestedPage))

        self.conn.send(response)
        self.conn.close()

    def getRoute(self):
        route = self.parsedHeaders["route"]
        if route in self.config["redirects"].keys():
            route = self.config["redirects"][route]
        path = os.path.join(self.config["fileslocation"], route)
        if self.config["scripting"]:
            routeContents = self.readFile(path, False)
            if routeContents:
                local = {}
                pattern = re.compile(r"(<py>(.*?)</py>)", flags=(re.S | re.M | re.I))
                for (m1,m2) in re.findall(pattern, routeContents):
                    exec(m2.strip(),{"headers":self.parsedHeaders},local)
                    if "res" in local.keys():
                        routeContents = routeContents.replace(m1, str(local["res"]) )
                    local["res"] = None
                return routeContents.encode()
            else:
                return None
        else:
            return self.readFile(path, True)

    def readFile(self, file, raw=True):
        try:
            with open(file, "rb") as fileHandle:
                file = fileHandle.read()
                if raw:
                    return file
                else:
                    return file.decode()
        except:
            return None

    def logMessage(self, message, error=False, fatal=False):
        formattedMessage = "[{}][{}] {}\n".format(error and "ERROR" or "MESSAGE", self.timeNow, message)
        print(formattedMessage.strip())
        if fatal and not self.config["logging"]["onlyfatal"]:
            try:
                with open(self.config["logging"]["logfile"],"a") as fileHandle:
                    fileHandle.write(formattedMessage)
            except:
                print("Failed to log to file.")
        if fatal: exit()

    def genHeaders(self, responseCode):
        return f"HTTP/1.1 {responseCode}\nDate: {self.timeNow()}\nServer: DSS\nConnection: close\n\n".encode()

    def parseHeaders(self, data):
        headersSplit = data.split("\r\n")
        routeSplit = headersSplit[0].split(" ")
        self.parsedHeaders["method"] = routeSplit[0]
        self.parsedHeaders["route"] = routeSplit[1]
        if self.config["scripting"]: # we only really need the other headers if we are scripting...
            for header in headersSplit[1:]:
                if header: # Dont do this....
                    headerSplit = header.split(":")
                    self.parsedHeaders[headerSplit[0].lower().replace('-',"_")] = headerSplit[1].strip()

    def timeNow(self):
        return time.strftime("%a, %d %b %Y %H:%M:%S", time.localtime())

    def clientLoop(self):
        self.conn, self.addr = self.s.accept()
        self.handleClient()

ws = webserver("config.json")

while True:
    ws.clientLoop()
			